src/java/arjdbc/jdbc/RubyJdbcConnection.java

```diff
@@ -2492,56 +2992,85 @@ protected void setObjectParameter(final ThreadContext context,
         statement.setObject(index, value.toJava(Object.class));
     }
 
-    protected Connection getConnection(ThreadContext context, boolean reportMissingConnection) {
-        Connection connection = (Connection) dataGetStruct(); // synchronized
+    /**
+     * Always returns a connection (might cause a reconnect if there's none).
+     * @return connection
+     * @throws <code>ActiveRecord::ConnectionNotEstablished</code>, <code>ActiveRecord::JDBCError</code>
+     */
+    protected final Connection getConnection() throws RaiseException {
+        return getConnection(false);
+    }
 
-        if (connection == null && reportMissingConnection) {
-            throw new RaiseException(getRuntime(), ActiveRecord(context).getClass("ConnectionNotEstablished"),
-                    "no connection available", false);
+    /**
+     * @see #getConnection()
+     * @param required set to true if a connection is required to exists (e.g. on commit)
+     * @return connection
+     * @throws <code>ActiveRecord::ConnectionNotEstablished</code> if disconnected
+     * @throws <code>ActiveRecord::JDBCError</code> if not connected and connecting fails with a SQL exception
+     */
+    protected final Connection getConnection(final boolean required) throws RaiseException {
+        try {
+            return getConnectionInternal(required);
+        }
+        catch (SQLException e) {
+            throw wrapException(getRuntime().getCurrentContext(), e);
         }
+    }
 
+    private Connection getConnectionInternal(final boolean required) throws SQLException {
+        Connection connection = getConnectionImpl();
+        if ( connection == null ) {
+            if ( required && ! connected ) {
+                final Ruby runtime = getRuntime();
+                final RubyClass errorClass = getConnectionNotEstablished( runtime );
+                throw new RaiseException(runtime, errorClass, "no connection available", false);
+            }
+            synchronized (this) {
+                connection = getConnectionImpl();
+                if ( connection == null ) {
+                    connectImpl( true ); // throws SQLException
+                    connection = getConnectionImpl();
+                }
+            }
+        }
         return connection;
     }
 
-    private IRubyObject setConnection(ThreadContext context, final Connection connection) {
-        close(getConnection(context, false)); // close previously open connection if there is one
-
-        final IRubyObject rubyConnectionObject = connection != null ?
-                convertJavaToRuby(connection) : context.runtime.getNil();
+    /**
+     * @note might return null if connection is lazy
+     * @return current JDBC connection
+     */
+    protected final Connection getConnectionImpl() {
```

Comment:
another final method.

Reply:
this one, I believe, should be fine - there's `getConnection` to override.
which than calls this piece - but this is a bit sensitive impl detail. 
makes no sense to override at this point.

Reply:
ok